skpro/distributions/qpd.py

```diff
@@ -0,0 +1,245 @@
+    qv_high : float or array_like[float]
+        quantile function value of quantile ``1 - alpha``
+    l : float
+        lower bound of semi-bounded range (default is 0)
+    version: str
+        options are ``normal`` (default) or ``logistic``
+    """
+
+    _tags = {
+        "capabilities:approx": ["pdfnorm"],
+        "capabilities:exact": ["mean", "var", "cdf", "ppf"],
+        "distr:measuretype": "continuous",
+    }
+
+    def __init__(
+        self,
+        alpha: float,
+        qv_low: float or object,
+        qv_median: float or object,
+        qv_high: float or object,
+        l: Optional[float] = 0,
+        version: Optional[str] = "normal",
+        index=None,
+        columns=None,
+    ):
+        self.alpha = alpha
+        self.version = version
+        self.qv_low = qv_low
+        self.qv_median = qv_median
+        self.qv_high = qv_high
+        self.index = index
+        self.columns = columns
+
+        for qv in [qv_low, qv_median, qv_high]:
+            if isinstance(qv, float):
+                qv = np.array([qv])
+            elif (
+                isinstance(qv, tuple)
+                or isinstance(qv, list)
+                or isinstance(qv, np.ndarray)
+            ):
+                qv = np.array(qv)
+            else:
+                raise ValueError("data type is not float or array_like object")
+
+        if version == "normal":
+            self.phi = norm()
+        elif version == "logistic":
+            self.phi = logistic()
+        else:
+            raise Exception("Invalid version.")
+
+        if (np.any(qv_low > qv_median)) or np.any((qv_high < qv_median)):
+            warnings.warn(
+                "The SPT values are not monotonically increasing, each SPT will be replaced by mean value"
+            )
+            idx = np.where((qv_low > qv_median), True, False) + np.where(
+                (qv_high < qv_median), True, False
+            )
+            print(f"replaced index by mean {np.argwhere(idx > 0).tolist()}")
+            qv_low[idx] = np.nanmean(qv_low)
+            qv_median[idx] = np.nanmean(qv_median)
+            qv_high[idx] = np.nanmean(qv_high)
+
+        self.l = l
+        self.c = self.phi.ppf(1 - alpha)
+
+        self.L = log(qv_low - l)
+        self.H = log(qv_high - l)
+        self.B = log(qv_median - l)
+
+        self.n = np.zeros_like(self.L)
+        self.theta = qv_median - l
+
+        pos = np.where((self.L + self.H - 2 * self.B) > 0, True, False)
+        self.n[pos] = 1
+        self.theta[pos] = qv_low[pos] - l
+
+        neg = np.where((self.L + self.H - 2 * self.B) < 0, True, False)
+        self.n[neg] = -1
+        self.theta[neg] = qv_high[neg] - l
+
+        B_L = self.B - self.L
+        H_B = self.H - self.B
+        _min = np.where(B_L < H_B, B_L, H_B)
+
+        self.delta = 1.0 / self.c * sinh(arccosh((self.H - self.L) / (2 * _min)))
+        self.kappa = 1.0 / (self.delta * self.c) * _min
+
+        # DataFrame
```

Comment:
I don't get this index thing here.

Reply:
yes, this does not look right to me. The index etc usually only "lives" in the distribution itself, not in the parameters.